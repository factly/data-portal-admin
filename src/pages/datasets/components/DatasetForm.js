import React, { useState } from 'react';
import { Form, Input, InputNumber, Button, Select, notification } from 'antd';

import Selector from '../../../components/Selector';
import MediaUploader from '../../media/components/MediaUpload';

const { Option } = Select;

const formItemLayout = {
  labelCol: { span: 6 },
  wrapperCol: { span: 14 },
};

const DatasetForm = ({ onSubmit, setDatasetId, data, next }) => {
  const [media, setMedia] = useState(data?.feature_media);

  const formFields = [
    { label: 'Title', name: 'title', placeholder: 'Dataset title', required: true },
    {
      label: 'Contact Email',
      name: 'contact_email',
      placeholder: 'shashi@factly.in',
      required: true,
      type: 'email',
    },
    { label: 'Contact Name', name: 'contact_name', placeholder: 'Shashi', required: true },
    { label: 'Description', name: 'description', placeholder: 'description' },
    { label: 'Data Standard', name: 'data_standard', placeholder: 'standard' },
    { label: 'Price', type: 'number', name: 'price', placeholder: '1000' },
    { label: 'Currency', name: 'currency_id', placeholder: 'INR' },
    { label: 'Frequency', name: 'frequency', placeholder: '1 month', required: true },
    { label: 'Granularity', name: 'granularity', placeholder: 'granularity' },
    { label: 'License', name: 'license', placeholder: 'MIT License' },
    {
      label: 'Related Articles',
      name: 'related_articles',
      placeholder: 'link://to.related.articles',
    },
    { label: 'Source', name: 'source', placeholder: 'link://to.source.com' },
    { label: 'Temporal Coverage', name: 'temporal_coverage', placeholder: 'temporal_coverage' },
    { label: 'Time Saved', type: 'number', name: 'time_saved', placeholder: '12' },
  ];

  const onUploadSuccess = (media) => {
    setMedia(media);
  };

  const onFinish = (values) => {
    values.frequency = `${values.frequency.count} ${values.frequency.units}`;
    if (media) values.featured_media_id = media.id;

    onSubmit(values)
      .then((data) => {
        setDatasetId(data.id);
        notification.success({
          message: 'Success',
        });
        next();
      })
      .catch((error) => {
        console.log(error);
        notification.error({
          message: 'Error',
          description: 'Something went wrong',
        });
      });
  };

  return (
    <Form
      name="datasets_create_step_one"
      initialValues={data}
      {...formItemLayout}
      onFinish={onFinish}
    >
      {formFields.map((field) => (
        <Form.Item
          key={field.name}
          label={field.label}
          name={field.name}
          rules={
            field.required && [
              {
                required: true,
                message: `Please enter ${field.label}!`,
              },
            ]
          }
        >
          {field.type === 'number' ? (
            <InputNumber placeholder={field.placeholder} />
          ) : field.name === 'currency_id' ? (
            <Selector action="Currencies" field="iso_code" />
          ) : field.name === 'frequency' ? (
            <Input.Group compact>
              <Form.Item
                name={['frequency', 'count']}
                noStyle
                rules={[{ required: true, message: 'Frequency count is required' }]}
              >
                <InputNumber min={1} style={{ width: '15%' }} placeholder="12" />
              </Form.Item>
              <Form.Item
                name={['frequency', 'units']}
                noStyle
                rules={[{ required: true, message: 'Frequency unit is required' }]}
              >
                <Select placeholder="Select Unit">
                  <Option value="days">Days</Option>
                  <Option value="months">Months</Option>
                  <Option value="years">Years</Option>
                </Select>
              </Form.Item>
            </Input.Group>
          ) : (
            <Input placeholder={field.placeholder} type={field.type || ''} />
          )}
        </Form.Item>
      ))}
      <Form.Item key={'medium'} label={'Medium'}>
        {/* {media && media.name} */}
        <MediaUploader onUploadSuccess={onUploadSuccess} />
      </Form.Item>
      <Form.Item wrapperCol={{ span: 12, offset: 6 }}>
        <Button type="primary" htmlType="submit">
          Submit
        </Button>
      </Form.Item>
    </Form>
  );
};

export default DatasetForm;
