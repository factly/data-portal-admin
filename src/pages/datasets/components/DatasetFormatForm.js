import React from 'react';
import { notification } from 'antd';
import { useDispatch, useSelector } from 'react-redux';

import { loadFormats } from '../../../actions/formats';
import Uploader from '../../../components/Uploader';
import DatasetFormatsList from './DatasetFormatsList';

const DatasetFormatForm = ({ datasetFormats, datasetId, onSubmit, prev }) => {
  const dispatch = useDispatch();
  const formats = useSelector(({ formats }) => Object.values(formats.items));

  React.useEffect(() => {
    dispatch(loadFormats());
  }, []);

  const onFinish = (values) => {
    try {
      Object.values(values).map((file) => {
        const data = { url: file.uploadURL };
        const index = formats
          .map((format) => format.name.toLowerCase())
          .indexOf(file.extension.toLowerCase());

        if (index !== -1) {
          data.format_id = formats[index].id;
        }

        onSubmit(data)
          .then(() => {
            notification.success({
              message: 'Success',
              description: 'Dataset Format succesfully added',
            });
          })
          .catch((error) => {
            console.log(error);

            notification.error({
              message: 'Error',
              description: 'Something went wrong',
            });
          });
      });
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <>
      <Uploader onUploadSuccess={onFinish} />
      <DatasetFormatsList datasetId={datasetId} datasetFormats={datasetFormats} />
    </>
  );
};

export default DatasetFormatForm;
