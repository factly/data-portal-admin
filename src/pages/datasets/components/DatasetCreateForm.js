import React, { useState } from 'react';
import { Steps } from 'antd';

import DatasetForm from './DatasetForm';
import DatasetFormatForm from './DatasetFormatForm';

const { Step } = Steps;

const DatasetCreateForm = ({ dataset, datasetFormats, onSubmitDataset, onSubmitDatasetFormat }) => {
  const [currentStep, setCurrentStep] = useState(0);
  const [datasetId, setDatasetId] = useState(dataset?.id);

  const next = () => {
    setCurrentStep(currentStep + 1);
  };
  const prev = () => {
    setCurrentStep(currentStep - 1);
  };

  const steps = [
    {
      title: 'Add Dataset Details',
      content: (
        <DatasetForm
          data={dataset}
          onSubmit={onSubmitDataset}
          setDatasetId={setDatasetId}
          next={next}
        />
      ),
    },
    {
      title: 'Add Datasets',
      content: (
        <DatasetFormatForm
          datasetFormats={datasetFormats}
          datasetId={datasetId}
          onSubmit={(data) => onSubmitDatasetFormat(datasetId, data)}
          prev={prev}
        />
      ),
    },
  ];

  return (
    <>
      <Steps current={currentStep} size="small">
        {steps.map((item) => (
          <Step key={item.title} title={item.title} />
        ))}
      </Steps>
      <br />
      {steps[currentStep].content}
    </>
  );
};

export default DatasetCreateForm;
