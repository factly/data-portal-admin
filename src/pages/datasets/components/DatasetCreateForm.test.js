import React from 'react';
import { useDispatch } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { act } from '@testing-library/react';
import { Steps } from 'antd';

import '../../../matchMedia.mock';
import DatasetCreateForm from './DatasetCreateForm';
import DatasetForm from './DatasetForm';
import DatasetFormatForm from './DatasetFormatForm';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

const dataset = { id: 1, title: 'title' };
const datasetFormat = {
  id: 1,
  url: 'url',
  format: { name: 'format' },
  created_at: '2020-12-12',
};

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('DatasetCreateForm', () => {
  const props = {
    dataset: dataset,
    datasetFormats: [datasetFormat],
    onSubmitDataset: jest.fn(),
    onSubmitDatasetFormat: jest.fn(),
  };

  const store = mockStore({});
  store.dispatch = jest.fn();
  useDispatch.mockReturnValue(jest.fn());

  describe('snapshot testing', () => {
    let wrapper;
    afterEach(() => {
      act(() => {
        wrapper.unmount();
      });
    });
    it('should render the component', () => {
      act(() => {
        wrapper = shallow(<DatasetCreateForm />);
      });

      expect(wrapper).toMatchSnapshot();
    });
    it('should match component in all steps', () => {
      act(() => {
        wrapper = shallow(<DatasetCreateForm {...props} />);
      });

      expect(wrapper).toMatchSnapshot();
      act(() => {
        wrapper.find(DatasetForm).props().next();
      });
      expect(wrapper.find(Steps).props().current).toEqual(1);
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    beforeEach(() => {
      act(() => {
        wrapper = shallow(<DatasetCreateForm {...props} />);
      });
    });
    afterEach(() => {
      act(() => {
        wrapper.unmount();
      });
    });
    it('should render the component with proper props', () => {
      const datasetFormProps = wrapper.find(DatasetForm).props();
      expect(datasetFormProps.data).toEqual(dataset);
      act(() => {
        wrapper.find(DatasetForm).props().next();
      });
      wrapper.update();
      const datasetFormatFormProps = wrapper.find(DatasetFormatForm).props();
      expect(datasetFormatFormProps.datasetFormats).toEqual([datasetFormat]);
      expect(datasetFormatFormProps.datasetId).toEqual(1);
    });
    it('should call setDatasetId', () => {
      wrapper.find(DatasetForm).props().setDatasetId(2);
      act(() => {
        wrapper.find(DatasetForm).props().next();
      });
      wrapper.update();
      expect(wrapper.find(DatasetFormatForm).props().datasetId).toEqual(2);
    });
    it('should call onSubmitDataset', () => {
      wrapper.find(DatasetForm).props().onSubmit({ test: 'test' });
      expect(props.onSubmitDataset).toHaveBeenCalledWith({ test: 'test' });
    });
    it('should call onSubmitDatasetFormat', () => {
      act(() => {
        wrapper.find(DatasetForm).props().next();
      });
      wrapper.update();
      wrapper.find(DatasetFormatForm).props().onSubmit({ test: 'test' });
      expect(props.onSubmitDatasetFormat).toHaveBeenCalledWith(1, { test: 'test' });
    });
    it('should handle next and prev buttons', () => {
      act(() => {
        wrapper.find(DatasetForm).props().next();
      });
      wrapper.update();
      expect(wrapper.find(Steps).props().current).toEqual(1);

      act(() => {
        wrapper.find(DatasetFormatForm).props().prev();
      });
      wrapper.update();
      expect(wrapper.find(Steps).props().current).toEqual(0);
    });
  });
});
