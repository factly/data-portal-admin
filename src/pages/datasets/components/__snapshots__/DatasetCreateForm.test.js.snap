// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetCreateForm snapshot testing should match component in all steps 1`] = `
<Fragment>
  <Steps
    current={0}
    size="small"
  >
    <Step
      key="Create Dataset"
      title="Create Dataset"
    />
    <Step
      key="Add Formats"
      title="Add Formats"
    />
  </Steps>
  <br />
  <DatasetForm
    data={
      Object {
        "id": 1,
        "title": "title",
      }
    }
    next={[Function]}
    onSubmit={[MockFunction]}
    setDatasetId={[Function]}
  />
</Fragment>
`;

exports[`DatasetCreateForm snapshot testing should match component in all steps 2`] = `
<Fragment>
  <Steps
    current={1}
    size="small"
  >
    <Step
      key="Create Dataset"
      title="Create Dataset"
    />
    <Step
      key="Add Formats"
      title="Add Formats"
    />
  </Steps>
  <br />
  <DatasetFormatForm
    datasetFormats={
      Array [
        Object {
          "created_at": "2020-12-12",
          "format": Object {
            "name": "format",
          },
          "id": 1,
          "url": "url",
        },
      ]
    }
    datasetId={1}
    onSubmit={[Function]}
    prev={[Function]}
  />
</Fragment>
`;

exports[`DatasetCreateForm snapshot testing should render the component 1`] = `
<Fragment>
  <Steps
    current={0}
    size="small"
  >
    <Step
      key="Create Dataset"
      title="Create Dataset"
    />
    <Step
      key="Add Formats"
      title="Add Formats"
    />
  </Steps>
  <br />
  <DatasetForm
    next={[Function]}
    setDatasetId={[Function]}
  />
</Fragment>
`;
